clusterloader2/pkg/test/simpletestexecutor.go

```diff
@@ -27,21 +35,164 @@ func createSimpleTestExecutor() TestExecutor {
 }
 
 // ExecuteTest executes test based on provided configuration.
-func (ste *simpleTestExecutor) ExecuteTest(ctx Context, conf *api.Config) error {
-	return nil
+func (ste *simpleTestExecutor) ExecuteTest(ctx Context, conf *api.Config) []error {
+	defer cleanupResources(ctx)
+	var errList []error
+	err := ctx.GetFramework().CreateAutomanagedNamespaces(int(conf.AutomanagedNamespaces))
+	if err != nil {
+		errList = append(errList, fmt.Errorf("automanaged namespaces creation failed: %v", err))
+		return errList
+	}
+
+	for i := range conf.Steps {
+		if stepErrList := ste.ExecuteStep(ctx, &conf.Steps[i]); len(stepErrList) > 0 {
+			errList = append(errList, stepErrList...)
+			if isErrsCritical(stepErrList) {
+				return errList
+			}
+		}
+	}
+	return errList
 }
 
 // ExecuteStep executes single test step based on provided step configuration.
-func (ste *simpleTestExecutor) ExecuteStep(ctx Context, step *api.Step) error {
-	return nil
+func (ste *simpleTestExecutor) ExecuteStep(ctx Context, step *api.Step) []error {
+	var errList []error
+	if len(step.Measurements) > 0 {
+		// TODO: handle measurments
+	} else {
+		var wg wait.Group
+		var lock sync.Mutex
+		for i := range step.Phases {
+			phase := &step.Phases[i]
+			wg.Start(func() {
+				if phaseErrList := ste.ExecutePhase(ctx, phase); len(phaseErrList) > 0 {
+					lock.Lock()
+					defer lock.Unlock()
+					errList = append(errList, phaseErrList...)
+				}
+			})
+		}
+		wg.Wait()
+	}
+	return errList
 }
 
 // ExecutePhase executes single test phase based on provided phase configuration.
-func (ste *simpleTestExecutor) ExecutePhase(ctx Context, phase *api.Phase) error {
+func (ste *simpleTestExecutor) ExecutePhase(ctx Context, phase *api.Phase) []error {
+	var errList []error
+	nsList := createNamespacesList(phase.NamespaceRange)
+	for _, nsName := range nsList {
+		instancesStates := make([]*state.InstancesState, 0)
+		for j := range phase.ObjectBundle {
+			id, err := getIdentifier(&phase.ObjectBundle[j])
+			if err != nil {
+				errList = append(errList, err)
+				return errList
+			}
+			instances, exists := ctx.GetState().Get(nsName, id)
+			if !exists {
+				instances = &state.InstancesState{
+					DesiredReplicaCount: 0,
+					CurrentReplicaCount: 0,
+					Object:              phase.ObjectBundle[j],
+				}
+			}
+			instances.DesiredReplicaCount = phase.ReplicasPerNamespace
+			ctx.GetState().Set(nsName, id, instances)
+			instancesStates = append(instancesStates, instances)
+		}
+
+		var maxCurrentReplicaCount int32
+		for j := range instancesStates {
+			if instancesStates[j].CurrentReplicaCount > maxCurrentReplicaCount {
+				maxCurrentReplicaCount = instancesStates[j].CurrentReplicaCount
```

Comment:
Could you explain what you're trying to do here?

Reply:
This the handling for the situation where we have uneven number of object instances:
Let's assume that in cluster we have [A,B,C]x10 and  [A,B]x10. Then we got phase with [A,B,C] and replicasPerNamespace = 0. In this case we want to delete every bundle using [C -> B -> A] order. Because number of instances is different for different objects, we have to calculate max(|{A}|,|{B}|, |{C}| ) first.

Reply:
Ok, now I understand. I think that sounds reasonable.

Could you add a brief comment above each code block in this function so it's easier to understand?

Reply:
done.